extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPusher.java

```diff
@@ -180,4 +169,33 @@ public DataSegment call() throws Exception
         config.isUseS3aSchema() ? "s3a" : "s3n"
     );
   }
+
+  private void putObject(String bucketName, String path, S3Object object, boolean replaceExisting)
+      throws S3ServiceException
+  {
+    object.setBucketName(bucketName);
+    object.setKey(path);
+    if (!config.getDisableAcl()) {
+      object.setAcl(GSAccessControlList.REST_CANNED_BUCKET_OWNER_FULL_CONTROL);
+    }
+
+    log.info("Pushing %s.", object);
+
+    if (!replaceExisting && doesObjectExist(bucketName, object.getKey())) {
+      log.info("Skipping push because key [%s] exists && replaceExisting == false", object.getKey());
+    } else {
+      s3Client.putObject(bucketName, object);
+    }
+  }
+
+  private boolean doesObjectExist(String bucketName, String objectKey)
+  {
+    try {
+      s3Client.getObjectDetails(bucketName, objectKey);
+      return true;
+    }
+    catch (ServiceException e) {
```

Comment:
This should be `S3Utils.isObjectInBucket(s3Client, bucketName, objectKey)`, which also does more fine grained error checking.

Reply:
Awesome, that's what I was looking for, thanks!